### PR TITLE
Add GPU Thomas solver

### PR DESCRIPTION
**Pull request by sriharshakandala**
- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: examples/hybrid/schur_complement_W.jl [hunk 1]
```diff
@@ -206,3 +201,90 @@ function linsolve!(::Type{Val{:init}}, f, u0; kwargs...)
...
+    return nothing
+end
+
+function column_thomas_solve!(::ClimaComms.CPUSingleThreaded, A, xᶠ𝕄, dims)
+    (Ni, Nj, Nh) = dims
+    (; S, S_column_array) = A
+    for h in 1:Nh, j in 1:Nj, i in 1:Ni
+        xᶠ𝕄_column_view = parent(Spaces.column(xᶠ𝕄, i, j, h))
+        S_column = Spaces.column(S, i, j, h)
+        @views S_column_array.dl .= parent(S_column.coefs.:1)[2:end]
```
Comment (simonbyrne) [inline examples/hybrid/schur_complement_W.jl:240]: i'd like to be able to do this directly in place on `S_columns.coefs`

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

File: examples/hybrid/schur_complement_W.jl [hunk 2]
```diff
@@ -206,3 +198,82 @@ function linsolve!(::Type{Val{:init}}, f, u0; kwargs...)
...
+    # back substitution
+    for row in (nrows - 1):-1:1
+        b[row] -= b[row + 1] * A[row, 3]
+    end
+    return nothing
+end
+
+function column_thomas_solve!(::ClimaComms.CPUSingleThreaded, A, b)
+    Ni, Nj, _, _, Nh = size(Spaces.local_geometry_data(axes(A)))
+    for h in 1:Nh, j in 1:Nj, i in 1:Ni
```
Comment (charleskawczynski) [inline examples/hybrid/schur_complement_W.jl:234]: Is there a reason we're not using `bycolumn` here? It would allow us to combine this and the multithreaded implementations

File: examples/hybrid/schur_complement_W.jl [hunk 3]
```diff
@@ -206,3 +198,82 @@ function linsolve!(::Type{Val{:init}}, f, u0; kwargs...)
...
+
+"""
+    thomas_algorithm!(A, b)
+
+Thomas algorithm for solving a linear system A x = b,
+where A is a tri-diagonal matrix.
+A and b are overwritten.
+Solution is written to b
+"""
+function thomas_algorithm!(A, b)
```
Comment (charleskawczynski) [inline examples/hybrid/schur_complement_W.jl:210]: Let's move this, and the remaining functions added in this file into `src/`, since we want it useable in ClimaAtmos.

File: perf/Project.toml [hunk 4]
```diff
@@ -1,6 +1,7 @@
 [deps]
 BenchmarkTools = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
 BlockArrays = "8e7c35d0-a365-5155-bbbb-fb81a777f24e"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
```
Comment (charleskawczynski) [inline perf/Project.toml:4]: What's this needed for? Maybe we can do this in a separate PR?

File: examples/hybrid/schur_complement_W.jl (hunk 3 above)
Comment (simonbyrne) [reply examples/hybrid/schur_complement_W.jl:210]: Yes, this should be part of ClimaCore itself.

File: examples/hybrid/schur_complement_W.jl [hunk 5]
```diff
@@ -206,3 +198,82 @@ function linsolve!(::Type{Val{:init}}, f, u0; kwargs...)
...
+    # A[_, 1] is lower diag, A[_, 2] is main diag, A[_, 3] is upper diag
+    nrows = size(A, 1)
+    # first row
+    A[1, 3] /= A[1, 2]
```
Comment (simonbyrne) [inline examples/hybrid/schur_complement_W.jl:214]: Accessing by `A[row, diag]`, is a bit confusing. Instead of accessing by `parent`, can we instead use `getidx` on the `Field`?

@dennisYatunin can you help with this?

Comment (dennisYatunin) [reply examples/hybrid/schur_complement_W.jl:214]: Since the matrix is represented by a field of matrix rows (`StencilCoefs`), we do have that `parent(A)[row, diag] == getidx(A, row, ...)[diag]`. However, there isn't a good way to use `setidx` to set an individual entry on a row; we can only use it to set all the entries in a row at once. This algorithm, as it is written, uses the third diagonal of `A` as a cache for computing `A \ b`, so it isn't really compatible with `setidx`. However, we could also allocate a cache instead of overwriting the third diagonal of `A`, and then use `setidx` to set the entries in that cache (each entry would just be a floating-point value). Would that be a good alternative?

Comment (dennisYatunin) [reply examples/hybrid/schur_complement_W.jl:214]: Also, we probably want to avoid modifying `A` in this function, because we might want to reuse the same Jacobian matrix for multiple Newton iterations. I think allocating a cache and leaving the third diagonal of `A` unchanged would be better than what we have here. Note that the current code doesn't modify the original matrix field; instead, it copies the field data into an array, and then modifies the third diagonal in that array.

Comment (simonbyrne) [reply examples/hybrid/schur_complement_W.jl:214]: No, I would prefer that we don't allocate a cache for now: the extra data movement will start to get expensive. If we need it, we can add it as an option.

Comment (charleskawczynski) [reply examples/hybrid/schur_complement_W.jl:214]: Can we clarify this: can we write this as `thomas_algorithm!(A, b, cache)`? so that `thomas_algorithm!` is non-allocating, but we have additional storage for handling things more intuitively?

File: examples/hybrid/test_thomas.jl [hunk 6]
```diff
@@ -0,0 +1,80 @@
+using Test
+using IntervalSets
+import Random: seed!
+import LinearAlgebra: Tridiagonal, norm
+import ClimaComms
+import ClimaCore:
+    Geometry, Domains, Meshes, Topologies, Spaces, Fields, Operators
+
+include("schur_complement_W.jl")
```
Comment (charleskawczynski) [inline examples/hybrid/test_thomas.jl:9]:
```suggestion
```

File: examples/hybrid/schur_complement_W.jl (hunk 5 above)
Comment (charleskawczynski) [reply examples/hybrid/schur_complement_W.jl:214]: Also, I'm not a fan of overwriting `b` with `x`. It's very awkward from a user POV. We have to compute tendencies, pass them in, and suddenly it becomes the state vector. If, for some reason, we want to keep the tendencies around (say for computing the residual, or for limiters or something), then this function corrupts that data. One extra field is not a performance issue here.

Comment (simonbyrne) [reply examples/hybrid/schur_complement_W.jl:214]: This isn't trying to be a generic `ldiv!` (which is why I gave it a different name).

Look, there are lots of ways we can do this. For now, just make it completely in-place, with no cache. That solves the immediate problem in the simplest way possible. If we decide we need something else, we can do that then.

Comment (simonbyrne) [reply examples/hybrid/schur_complement_W.jl:214]: As for the `getidx`/`setidx!`, you can define
```
ld = S.coefs.:1
d = S.coefs.:2
ud = S.coefs.:3
```
and operate on those

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

File: src/Operators/thomas_algorithm.jl [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    end
+end
+
+# This is the same as @inbounds Fields.field_values(column_field)[index]
+_getindex(column_field, index) = @inbounds Operators.getidx(
+    axes(column_field),
+    column_field,
+    Operators.Interior(),
+    index - 1 + Operators.left_idx(axes(column_field)),
+)
```
Comment (simonbyrne) [inline src/Operators/thomas_algorithm.jl:80]: why not just used `getidx` directly?

Comment (dennisYatunin) [reply src/Operators/thomas_algorithm.jl:80]: I just though it would be a bit hard to read with all the `space`s and `loc`s and `- 1 + left_idx`s floating around everywhere.

File: src/Operators/thomas_algorithm.jl [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    idx = threadIdx().x + (blockIdx().x - 1) * blockDim().x
+    Ni, Nj, _, _, Nh = size(Fields.field_values(A))
+    if idx <= Ni * Nj * Nh
+        i, j, h = Spaces._get_idx((Ni, Nj, Nh), idx)
+        thomas_algorithm!(Spaces.column(A, i, j, h), Spaces.column(b, i, j, h))
+    end
+    return nothing
+end
+
+function thomas_algorithm!(A, b)
```
Comment (charleskawczynski) [inline src/Operators/thomas_algorithm.jl:33]:
```suggestion
function thomas_algorithm!(A::Fields.ColumnField, b::Fields.ColumnField)
```

File: src/Operators/thomas_algorithm.jl [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+column_thomas_solve!(::ClimaComms.AbstractCPUDevice, A, b) =
+    Fields.bycolumn(colidx -> thomas_algorithm!(A[colidx], b[colidx]), axes(A))
+
+function column_thomas_solve!(::ClimaComms.CUDADevice, A, b)
+    Ni, Nj, _, _, Nh = size(Fields.field_values(A))
+    nthreads, nblocks = Spaces._configure_threadblock(Ni * Nj * Nh)
+    @cuda threads = nthreads blocks = nblocks thomas_algorithm_kernel!(A, b)
+end
+
+function thomas_algorithm_kernel!(A, b)
```
Comment (charleskawczynski) [inline src/Operators/thomas_algorithm.jl:23]:
```suggestion
thomas_algorithm_kernel!(A::Fields.ColumnField, b::Fields.ColumnField) = thomas_algorithm!(A, b)
function thomas_algorithm_kernel!(A::Fields.ExtrudedFiniteDifferenceField, b::Fields.ExtrudedFiniteDifferenceField)
```

**Review by charleskawczynski — CHANGES_REQUESTED** [review]
We need to make `column_thomas_solve!` work for `Fields.ExtrudedFiniteDifferenceField` and `Fields.ColumnField`. Let's handle this with dispatch (let's add types for these methods so it's clear what types of fields they're valid for, and so that we get better error messages), we'll need an extra method or two. Let me know if you need help.

**Comment by dennisYatunin** [comment]
bors try

**Comment by dennisYatunin** [comment]
bors try

**Review by charleskawczynski — COMMENTED** [review]
Unfortunately, nothing seems to be converging, so there seems to be a translational bug somewhere. Maybe we can add a test against the original code to identify the issue?

File: test/Operators/thomas_algorithm.jl [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    Ni, Nj, _, _, Nh = size(Fields.field_values(A))
+    for i in 1:Ni, j in 1:Nj, h in 1:Nh
+        A_column_data = Array(parent(Spaces.column(A_copy, i, j, h)))
+        A_column_array = Tridiagonal(
+            A_column_data[2:end, 1],
+            A_column_data[:, 2],
+            A_column_data[1:(end - 1), 3],
+        )
+        b_column_array = Array(parent(Spaces.column(b_copy, i, j, h)))[:]
+        x_column_array = Array(parent(Spaces.column(b, i, j, h)))[:]
```
Comment (charleskawczynski) [inline test/Operators/thomas_algorithm.jl:43]:
```suggestion
    Fields.bycolumn(axes(A)) do colidx
        A_column_data = Array(parent(A_copy[colidx]))
        A_column_array = Tridiagonal(
            A_column_data[2:end, 1],
            A_column_data[:, 2],
            A_column_data[1:(end - 1), 3],
        )
        b_column_array = Array(parent(b_copy[colidx]))[:]
        x_column_array = Array(parent(b[colidx]))[:]
```

Comment (dennisYatunin) [reply test/Operators/thomas_algorithm.jl:43]: I tried this earlier, and it doesn't work on the GPU

Comment (charleskawczynski) [reply test/Operators/thomas_algorithm.jl:43]: Why is `[:]` necessary here?

Comment (dennisYatunin) [reply test/Operators/thomas_algorithm.jl:43]: It's because the vector field is stored as a 5D array, and I need to reshape it into a 1D array. I think putting `[:]` at the end is equivalent to using the `vec` function; it creates a 1D view of the 5D parent array.

**Comment by charleskawczynski** [comment]
New build at https://buildkite.com/clima/climaatmos-ci/builds/10949

**Review by charleskawczynski — APPROVED** [review]
